aten/src/ATen/Dispatch.h

```diff
@@ -1,65 +1,71 @@
-    case ScalarType::Byte:
-      return F<uint8_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Char:
-      return F<int8_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Double:
-      return F<double>::apply(std::forward<Args>(args)...);
-    case ScalarType::Float:
-      return F<float>::apply(std::forward<Args>(args)...);
-    case ScalarType::Int:
-      return F<int>::apply(std::forward<Args>(args)...);
-    case ScalarType::Long:
-      return F<int64_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Short:
-      return F<int16_t>::apply(std::forward<Args>(args)...);
-    case ScalarType::Half:
-      return F<Half>::apply(std::forward<Args>(args)...);
-    default:
-      runtime_error("%s not implemented for '%s'", name, the_type.toString());
-    }
-}
+#define AT_DISPATCH_FLOATING_TYPES(TYPE, NAME, ...)                      \
+  [&] {                                                                  \
+    const at::Type& the_type = TYPE;                                     \
+    switch (the_type.scalarType()) {                                     \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Double, double, __VA_ARGS__)  \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Float, float, __VA_ARGS__)    \
+      default:                                                           \
+        at::runtime_error(                                               \
+            "%s not implemented for '%s'", (NAME), the_type.toString()); \
+    }                                                                    \
+  }()
 
-template<typename R, template <typename> class F, typename ... Args>
-R dispatch_floating_types(const Type& the_type, const char *name, Args&&... args) {
-  switch(the_type.scalarType()) {
-    case ScalarType::Double:
-      return F<double>::apply(std::forward<Args>(args)...);
-    case ScalarType::Float:
-      return F<float>::apply(std::forward<Args>(args)...);
-    case ScalarType::Half: // no native half math on either CPU or GPU.
-    default:
-      runtime_error("%s not implemented for '%s'", name, the_type.toString());
-  }
-}
+#define AT_DISPATCH_FLOATING_TYPES_AND_HALF(TYPE, NAME, ...)             \
+  [&] {                                                                  \
+    const at::Type& the_type = TYPE;                                     \
+    switch (the_type.scalarType()) {                                     \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Double, double, __VA_ARGS__)  \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Float, float, __VA_ARGS__)    \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Half, Half, __VA_ARGS__)      \
+      default:                                                           \
+        at::runtime_error(                                               \
+            "%s not implemented for '%s'", (NAME), the_type.toString()); \
+    }                                                                    \
+  }()
```

Comment:
Is there any benefit to defining the lambda and calling it immediately compared to just creating a new temporary scope (wrapping the code in `{ ... }`)?

Reply:
You can use `return` statements in a lambda, but not a temporary scope. This means that the dispatch functions can be used as expressions and the return type is inferred.